design/websocket_protocol.md

```diff
@@ -179,6 +201,191 @@ else if range.start < shifter's end caret:
         shifter's start caret = range.start
 
 if there was a replacement with a positive delta.doc:
-    do the same process ass the normal addition but with the editor's end caret
+    do the same process as the normal addition but with the editor's end caret
     equal to its start caret and delta.start instead of delta.doc
 ```
+
+### Message Protocol
+
+#### Server
+The server stores a checkpoint for each version of a conversation and removes a
+version checkpoint when it receives a `Sync` message from all active users in
+the conversation for the next version. The conversation version number is
+incremented with every valid `Edit Update` message received from a client. The
+checkpoint is used to handle `Cursor Update` messages from clients that are
+behind on the conversation state. Each checkpoint will contain the following
+fields:
+* `version`: int
+    * The conversation version number of this checkpoint.
+* `active_users`: Object
+    * User IDs are the keys and the value is another Object of the user's caret
+      `start` and `end` at this conversation version.
+* `sender_caret`: Object
+    * Caret position (`start` and `end`) of the sender of the
+      message that brought the conversation to this version.
+* `delta`: Object
+    * The deltas (`caret_start`, `caret_end`, and `doc`) that brought the
+      conversation to this version.
+
+The server will then follow the pseudocode below when processing `Update`,
+`Ack`, and `Sync` messages.
+
+```
+receive msg
+switch msg.type
+    case edit update:
+        apply msg.patch to conversation.content
+        if patch failed:
+            return
+        conversation.version++
+        if msg.version != conversation.version:
+            msg.version = conversation.version
+        broadcast message to all active clients, excluding the sender
+        sender_caret = current sender caret position
+        apply delta to caret of sender
+        shift carets of all active clients
```

Comment:
I think these two lines need to be switched because we need the sender's original caret when applying the shifting algorithm. It needs to be switched in the `patches` code too.
```suggestion
        shift carets of all active clients
        apply delta to caret of sender
```

Reply:
Good call.